lib/util/pthread_create_blocking.c

```diff
@@ -0,0 +1,279 @@
+			if (set_initialized(U) != 0)
+				*U->rc_sync_p = -1;
+			return (NULL);
+		}
+	}
+
+	/*
+	 * Set up the cleanup function.  If none was specified, use a "no-op"
+	 * function (see the comment for cleanup_fake()),
+	 */
+	if (cleanup_routine == NULL)
+		cleanup_routine = cleanup_fake;
+	pthread_cleanup_push(cleanup_routine, user_cookie);
+
+	/*
+	 * Indicate that the main thread can continue (which will free U), and
+	 * run the desired routine. */
+	if (set_initialized(U) != 0) {
+		*U->rc_sync_p = -1;
+		rc = NULL;
+	} else {
+		/* Run the desired routine. */
+		rc = run_routine(user_cookie);
+	}
+
+	/* Clean up. */
+	pthread_cleanup_pop(1);
+
+	/* Done. */
+	return (rc);
+}
+
+/* Wait up 1 second, or until the condition variable is signalled. */
+static int
+cond_timedwait(struct wrapped_cookie * U)
+{
+	struct timespec tp = {0};
+	struct timeval tv;
+	int rc;
+
+	/* 1 second later than the current time. */
+	if (monoclock_get(&tv)) {
+		warnp("monoclock_get");
+		goto err0;
+	}
+	tp.tv_sec = tv.tv_sec;
+	tp.tv_nsec = 1000 * tv.tv_usec;
+
+	/* Wait for up to 1 second. */
+	if ((rc = pthread_cond_timedwait(U->cond_p, U->mutex_p, &tp)) != 0) {
+		if (rc != ETIMEDOUT) {
+			warn0("pthread_cond_timedwait: %s", strerror(rc));
+			goto err0;
+		}
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failed! */
+	return (-1);
+}
+
+/**
+ * pthread_create_blocking(thread, attr, init_routine, run_routine,
+ *     cleanup_routine, arg):
+ * Run pthread_create() and block until the new ${thread} has started and
+ * ${init_routine} has finished.  ${init_routine} and ${cleanup_routine} may
+ * be NULL.  If ${init_routine} is not NULL and returns a non-zero value, do
+ * not attempt to run ${run_routine} and instead return that value.  When
+ * ${run_routine} finishes, make its returned value available via
+ * pthread_join().
+ */
+int
+pthread_create_blocking(pthread_t * restrict thread,
+    const pthread_attr_t * restrict attr, int (*init_routine)(void *),
+    void * (*run_routine)(void *), void (*cleanup_routine)(void *),
+    void * restrict arg)
+{
+	struct wrapped_cookie * U;
+	int rc = -1;
+	int rc_init = 0;
+	int rc_sync = 0;
+	int initialized = 0;
+	pthread_mutex_t mutex;
+	pthread_cond_t cond;
+
+	/* Allocate our cookie and assign user-specific values. */
+	if ((U = malloc(sizeof(struct wrapped_cookie))) == NULL) {
+		warnp("malloc");
+		goto err0;
+	}
+	U->user_cookie = arg;
+	U->init_routine = init_routine;
+	U->run_routine = run_routine;
+	U->cleanup_routine = cleanup_routine;
+
+	/* Initialize & assign thread startup-related variables. */
+	if ((rc = pthread_mutex_init(&mutex, NULL)) != 0) {
+		warn0("pthread_mutex_init: %s", strerror(rc));
+		goto err1;
+	}
+	if ((rc = pthread_cond_init(&cond, NULL)) != 0) {
+		warnp("pthread_cond_init: %s", strerror(rc));
+		goto err2;
+	}
+	U->mutex_p = &mutex;
+	U->cond_p = &cond;
+	U->initialized_p = &initialized;
+	U->rc_init_p = &rc_init;
+	U->rc_sync_p = &rc_sync;
```

Comment:
Why are you creating these variables on the stack and then passing pointers to them?  Wouldn't it be cleaner to just put them inside the structure?

Reply:
I had a reason.  I'm not saying it was a good reason.  I wanted to check `rc_init` after `free(U)`.  Other things grew out of that.

Your suggestion does sound cleaner, so I'll adjust accordingly.